client/remote.go

```diff
@@ -16,17 +16,103 @@ import (
 	"github.com/miekg/dns"
 )
 
+// connPoolType is a async safe pool of established gokeyless Conn
+// so we don't need to do TLS handshake unnecessarily.
+type connPoolType struct {
+	pool map[string]*Conn
+	sync.RWMutex
+}
+
+// connPool keeps all active Conn
+var connPool *connPoolType
+
 // A Remote represents some number of remote keyless server(s)
 type Remote interface {
-	Dial(*Client) (*gokeyless.Conn, error)
+	Dial(*Client) (*Conn, error)
 	Add(Remote) Remote
 }
 
+// A Conn represents a long-lived client connection to a keyserver.
+type Conn struct {
+	*gokeyless.Conn
+	addr string
+}
+
 // A singleRemote is an individual remote server
 type singleRemote struct {
-	net.Addr
-	ServerName string
-	conn       *gokeyless.Conn
+	net.Addr          // actual address
+	ServerName string // hostname for TLS verification
+}
+
+func init() {
+	connPool = &connPoolType{
+		pool: make(map[string]*Conn),
+	}
+}
+
+// NewConn creates a new Conn based on a gokeyless.Conn
+func NewConn(addr string, conn *gokeyless.Conn) *Conn {
+	c := Conn{
+		Conn: conn,
+		addr: addr,
+	}
+
+	go healthchecker(&c, 1*time.Second)
+	return &c
+}
+
+// Close closes a Conn and remove it from the conn pool
+func (conn *Conn) Close() {
+	conn.Conn.Close()
+	connPool.Remove(conn.addr)
+}
+
+// healthchecker is a recurrent timer function that tests the connections
+func healthchecker(conn *Conn, after time.Duration) {
+	for {
+		time.Sleep(after)
+		if !conn.Conn.IsClosed() {
+			err := conn.Ping(nil)
+			if err != nil {
+				log.Debug("health check ping failed:", err)
+				// shut down the conn and remove it from the
+				// conn pool.
+				conn.Close()
+				return
+			}
+			log.Debug("start a new health check timer")
+		}
+	}
+}
+
+// Get returns a Conn from the pool if there is any.
+func (p *connPoolType) Get(key string) *Conn {
+	p.RLock()
+	defer p.RUnlock()
+
+	conn, ok := p.pool[key]
```

Comment:
If ok is false, conn will be nil since it's the default value; you could just do 

```
conn := p.pool[key]
return conn
```

Assigning to a var and then returning the var should make sure that the map access occurs before the call to `Unlock`.